File: software/python_modules/tart/tart/util/angle.py
#
# Helper class for dealing with angles (and to avoid getting tangled between degrees and radians.
#
# Copyright (c) Tim Molteno 2013. tim@elec.ac.nz
#
import math
import numpy as np
from tart.util import constants

def deg_to_rad(x):
    return x * math.pi / 180.0

def wrap_2pi(x):
    r = math.fmod(x,constants.TWO_PI)
    if (r < 0.0):
        r += constants.TWO_PI
    return r

def wrap_360(x):
    r = math.fmod(x,360.0)
    if (r < 0.0):
        r += 360.0
    return r

def wrap_180(x):
    r = math.fmod(x,360.0)
    if (r > 180.0):
        r -= 360.0
    return r

# from functools import total_ordering
# @total_ordering
class Angle(object):

    def __init__(self, deg=0, minute=0, sec=0):
        dec_deg = deg + float(minute)/60.0 + float(sec) / 3600.0
        self.rad = deg_to_rad(dec_deg)

    def __repr__(self):
        return str(self.to_degrees())

    def to_elevation(self):
         deg = self.to_degrees()
         if (deg > 90.0):
             raise "Elevation out of RANGE %f" % deg
         if (deg < -90.0):
             raise "Elevation out of RANGE %f" % deg

         return from_dms(deg)

    def to_declination(self):
        return self.to_elevation()

    def to_ra(self):
        return from_rad(wrap_2pi(self.rad))

    def to_rad(self):
        return self.rad

    def to_degrees(self):
        return self.rad * (360.0 / constants.TWO_PI)

    def to_hours(self):
        return wrap_2pi(self.rad) * (24.0 / constants.TWO_PI)

    def to_dms(self):
        dd = self.to_degrees()
        mnt,sec = divmod(dd*3600,60) ## 60 seconds in a
        deg,mnt = divmod(mnt,60)
        return deg,mnt,sec

    def to_hms(self):
        dd = self.to_degrees()
        
        hour,rem = divmod(dd, 15)   # 15 Degrees in an hour
        mnt,sec = divmod(rem*3600/15,60)  # 60 Minutes in an hour
        return hour,mnt,sec

    def sin(self):
        return math.sin(self.rad)

    def cos(self):
        return math.cos(self.rad)

    def tan(self):
        return math.tan(self.rad)

    def __abs__(self):
        return from_rad(abs(self.rad))

    def __add__(self, x):
        return from_rad(self.rad + x.rad)

    def __radd__(self, x):
        return self.__add__(x)

    def __sub__(self, x):
        return from_rad(self.rad - x.rad)

    def __rsub__(self, x):
        return self.__sub__(x)

    def __eq__(self, x):
        return (wrap_2pi(self.rad) == wrap_2pi(x.rad))

    def __lt__(self, x):
        return self.rad < x.rad # TODO should we wrap both?

    def __neg__(self):
        return from_rad(-self.rad)

def from_hours(hour, minute=0.0, sec=0.0):
    hr = hour % 24.0
    return Angle(deg=hr*15, minute=minute*15, sec=sec*15)

def from_dms(deg=0.0, minute=0.0, sec=0.0):
    return Angle(deg, minute, sec)

def from_rad(rad):
    a = Angle(0.,0.,0.)
    a.rad = rad
    return a

def atan(x):
    return from_rad(np.arctan(x))

def atan2(x,y):
    return from_rad(np.arctan2(x,y))

def asin(x):
    return from_rad(np.arcsin(x))

def acos(x):
    return from_rad(np.arccos(x))

def add_rad(a, x):
    return from_rad(a.rad + x)

File: software/containers/ephemeris_server/code/Angle.rb
$LOAD_PATH.unshift File.dirname(__FILE__)
require 'Util'

class Angle

  DEG2RAD = Util::PI / 180.0
  RAD2DEG = 180.0 / Util::PI
  
  def initialize(radians)
    @theta = radians
  end

  def self.rad2deg(in_rad)
          return in_rad * RAD2DEG
  end

  def self.wrap360(in_rad)
      d = rad2deg(in_rad)
      d = d + 360.0 if d < 0
      return d
  end

  def self.deg2rad(in_deg)
  
          return in_deg * DEG2RAD
  end

  def self.from_deg(in_deg)
  
          return Angle.new(deg2rad(in_deg))
  end

  def self.from_rad(in_rad)
  
          return Angle.new(in_rad)
  end

  def self.from_dms(degrees, minutes, seconds)

          if (degrees < 0)
                  deg = degrees - minutes/60.0 - seconds/3600.0
          else
                  deg = degrees + minutes/60.0 + seconds/3600.0
          end

          return Angle.new(Angle.deg2rad(deg))
  end

  def radians()
    return @theta
  end

  def degrees()
    return Angle.rad2deg(@theta)
  end
  
  def cos()
    return Math.cos(@theta)
  end

  def to_s
    sign = "+"
    d = a.degrees()

    sign = "-" if (a.degrees() < 0) 
    d = -d if (a.degrees() < 0)
    
    deg = d.floor.to_i

    mind = (d - deg) * 60.0;
    minutes = mind.floor.to_i
    seconds = (mind - minutes) * 60.0;

    return "#{sign}#{deg} #{minutes}'#{seconds}\""
  end
end




